src/GafferScene/Shader.cpp

```diff
@@ -621,6 +621,19 @@ class Shader::NetworkBuilder
 					addParameterComponentConnections( it->get(), inputName, connections );
 				}
 			}
+			else
+			{
+				const Gaffer::Plug *effectiveParameter = this->effectiveParameter( parameter );
+				if( effectiveParameter && isOutputParameter( effectiveParameter ) )
+				{
+					IECore::InternedString inputName = parameterName.string();
+
+					connections.push_back( {
+						outputParameterForPlug( effectiveParameter ),
+						{ IECore::InternedString(), inputName }
+					} );
+				}
+			}
```

Comment:
It's not clear to me what this case is for - could definitely use a comment at least. It doesn't seem related to this PR's main purpose ( arrays, which are handled in the case above ). Maybe this should be obvious, but I can't think offhand of what data types have components, other than compoundNumerics, splines, and arrays?

Reply:
This was to allow support connections to the entire array or individual array indices.

Reply:
I'm still not quite following here - the conditional on line 615 should be handling arrays?

You're saying that if you make a connection to the entire array, then addParameterComponentConnections is called on a Plug that isn't an ArrayPlug?

Reply:
I have finally come back to this to implement John's change, and the reason for this block of code is to handle a float array that has a float plug connected to an index.

Reply:
Hmm, I'm still not 100% sure I understand this, but if I'm understanding correctly, you've identified a bug in the existing code, but I don't think this solution is in the right place. `addComponentConnections` is intended to add connections to the children of a plug, not the plug itself. I think the bug is on line 621 when the children of an array plug are handled just by calling addComponentConnections - this isn't right, because the children of an array might not have components ( ie. a float array ). But we don't exactly have the right code for "add a top level connection if the top level is connected, otherwise add component connections" in a form that's easy to call - if we just called addParameter on line 621, that would sort of do the trick, but it would also be trying to set the parameter value, which we don't want here. We could paste in the relevant code from addParameter, but that's a bit messy. Maybe @johnhaddon  has an idea how he'd like this arranged.

Reply:
It's not really a bug in current gaffer, it's very specific to array plugs as they have component connections which can either have sub component connections or just regular component connections. eg. `ramp['color'][1]['r']` vs `ramp['position'][1]`. This made the most sense to me as I didn't want to have a bunch of duplicate code looking for components within the array, I would be happy to move it anywhere it would make more sense.